lib/metasploit/framework/aws/client.rb

```diff
@@ -0,0 +1,173 @@
+
+          signature = hexhmac(credentials_mac, doc)
+          [headers_list, signature]
+        end
+
+        def auth(service, headers, body_digest, now)
+          headers_list, signature = sign(service, headers, body_digest, now)
+          "AWS4-HMAC-SHA256 Credential=#{datastore['ACCESS_KEY']}/#{now[0, 8]}/#{datastore['Region']}/#{service}/aws4_request, SignedHeaders=#{headers_list}, Signature=#{signature}"
+        end
+
+        def body(vars_post)
+          pstr = ""
+          vars_post.each_pair do |var,val|
+            pstr << '&' if pstr.length > 0
+            pstr << var
+            pstr << '='
+            pstr << val
+          end
+          pstr
+        end
+
+        def headers(service, body_digest, body_length)
+          now = Time.now.utc.strftime("%Y%m%dT%H%M%SZ")
+          headers = {
+            'Content-Type' => 'application/x-www-form-urlencoded; charset=utf-8',
+            'Accept-Encoding' => '',
+            'User-Agent' => "Metasploit #{Metasploit::Framework::VERSION} (jg)",
+            'X-Amz-Date' => now,
+            'Host' => datastore['RHOST'],
+            'X-Amz-Content-Sha256' => body_digest,
+            'Accept' => '*/*'
+          }
+          headers['X-Amz-Security-Token'] = datastore['TOKEN'] if datastore['TOKEN']
+          sign_headers = ['Content-Type', 'Host', 'User-Agent', 'X-Amz-Content-Sha256', 'X-Amz-Date']
+          auth_headers = headers.select { |k, _| sign_headers.include?(k) }
+          headers['Authorization'] = auth(service, auth_headers, body_digest, now)
+          headers
+        end
+
+        def print_hsh(hsh)
+          hsh.each do |key, value|
+            print_warning "#{key}: #{value}"
+          end
+        end
+
+        def print_results(doc, action)
+          response = "#{action}Response"
+          result = "#{action}Result"
+          resource = /[A-Z][a-z]+([A-Za-z]+)/.match(action)[1]
+
+          if doc["ErrorResponse"] && doc["ErrorResponse"]["Error"]
+            print_error doc["ErrorResponse"]["Error"]["Message"]
+            return nil
+          end
+
+          idoc = doc[response] if doc[response]
+          idoc = idoc[result] if idoc[result]
+          idoc = idoc[resource] if idoc[resource]
+
+          if idoc["member"]
+            idoc["member"].each do |x|
+              print_hsh x
+            end
+          else
+            print_hsh idoc
+          end
+          idoc
+        end
+
+        def call_api(service, api_params)
+          print_status("#{peer} - Connecting (#{datastore['RHOST']})...")
+          body = body(api_params)
+          body_length = body.length
+          body_digest = hexdigest(body)
+          res = send_request_raw(
+            'method' => 'POST',
+            'data' => body,
+            'headers' => headers(service, body_digest, body_length)
+          )
+          Hash.from_xml(res.body)
+        rescue => e
+          print_error e.message
```

Comment:
It may be worth updating any calls to `call_api` to handle the possibly null response.